libc-test/test/style/mod.rs

```diff
@@ -0,0 +1,465 @@
+//! Provides the [StyleChecker] visitor to verify the coding style of
+//! this library.
+//!
+//! This is split out so that the implementation itself can be tested
+//! separately, see test/check_style.rs for how it's used and
+//! test/style_tests.rs for the implementation tests.
+//!
+//! ## Guidelines
+//!
+//! The current style is:
+//!
+//! * Specific module layout:
+//!     1. use directives
+//!     2. typedefs
+//!     3. structs
+//!     4. constants
+//!     5. f! { ... } functions
+//!     6. extern functions
+//!     7. modules + pub use
+//! * No manual deriving Copy/Clone
+//! * Only one f! per module
+//! * Multiple s! macros are allowed as long as there isn't a duplicate cfg,
+//!   whether as a standalone attribute (#[cfg]) or in a cfg_if!
+//! * s! macros should not just have a positive cfg since they should
+//!   just go into the relevant file but combined cfgs with all(...) and
+//!   any(...) are allowed
+
+use std::collections::HashMap;
+use std::fs;
+use std::ops::Deref;
+use std::path::{Path, PathBuf};
+
+use annotate_snippets::{Level, Renderer, Snippet};
+use proc_macro2::Span;
+use syn::parse::{Parse, ParseStream};
+use syn::spanned::Spanned;
+use syn::visit::{self, Visit};
+use syn::Token;
+
+pub type Error = Box<dyn std::error::Error>;
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Default)]
+pub struct StyleChecker {
+    /// The state the style checker is in, used to enforce the module layout.
+    state: State,
+    /// Span of the first item encountered in this state to use in help
+    /// diagnostic text.
+    state_span: Option<Span>,
+    seen_s_macro_cfgs: HashMap<String, Span>,
+    /// Span of the first f! macro seen, used to enforce only one f! macro
+    /// per module.
+    first_f_macro: Option<Span>,
+    /// The errors that the style checker has seen.
+    errors: Vec<FileError>,
+    /// Path of the currently active file.
+    path: PathBuf,
+    /// Whether the style checker is currently in an `impl` block.
+    in_impl: bool,
+}
+
+#[derive(Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+enum State {
+    #[default]
+    Start,
+    Imports,
+    Typedefs,
+    Structs,
+    Constants,
+    FunctionDefinitions,
+    Functions,
+    Modules,
+}
+
+/// Similar to [syn::ExprIf] except with [syn::Attribute]
+/// as the condition instead of [syn::Expr].
+struct ExprCfgIf {
+    _attr: syn::Attribute,
+    then_branch: Vec<syn::Item>,
+    else_branch: Option<Box<ExprCfgElse>>,
+}
+
+enum ExprCfgElse {
+    Block(Vec<syn::Item>),
+    If(ExprCfgIf),
```

Comment:
Just some suggested docs

```suggestion
    /// Final block with no condition `else { /* ... */ }`.
    Block(Vec<syn::Item>),
    /// `else if { /* ... */ }` block.
    If(ExprCfgIf),
```